### ISSUE TITLE
pH7Builder v16, Release Candidate 2 is HERE???

### ISSUE THREAD

**Issue by Undefined-Variables** [pH7Software/pH7-Social-Dating-CMS#803]
@pH-7 why are we releasing new build variations all the time when we still have outstanding issues in the previous ones? This is part of the issue with the software. 

Before releasing new builds we should make sure the old one is perfect and its just not, there are issues in friends,  wall,install script inserts tables before checking if the field's values a correct resulting in failed admin account,  messages are outdated and the instant messenger is funny its so bad. 

We need to fix before  we evolve or we end up where we are now with problems spread out in almost every aspect of the software!

**Comment by polynamaude** [comment]
@Undefined-Variables A RC (Release Candidate) is somewhat a post Alpha/Beta release...Normally you freeze feature and only fix bugs on new stuff.
But like you said, I don't really get it at the way it's going on...

**Comment by pH-7** [comment]
Hiya! Great thought 👍 I appreciate you bring this up! 🙃

So, let’s dig in...
RC versions won’t bring any new major changes and certainly not new features. This is the reason why it’s called RC ans not alpha or beta versions. RC branches are the most stable and closest versions to the final release.
Some software companies/orgs out there would still bring major changes in a RC, but I believe this is wrong. 
An RC version is there just to make sure the new changes introduced in that particular version are working fine and didn’t break something else. And sometimes it can polish a little bit more the version too...

v16.0.0 has already been delayed for over one or two months, and the few new bugs you mentioned are already set for the new roadmap release (v16.1 and further), since they came lately.

And since v16.0.0 also includes some hot fixes (like the XSS vulnerability patch), the version (stable one) will now be released in one week max from now. 

I understand your concerns though.
Those issues you mentioned will be released in the next releases, in v16.1 and 16.2. And..., cherry on the cake, v17 will be a REALLY exciting one!!!

Thanks for bringing back those important issues, since indeed they have to be fixed and improved. You’re totally right.

By the way, I’m not aware of every of them as relevant bugs like the one during the installation. I would love if you could give me further details.


Thanks very much for your support! I highly appreciate it! 😊

**Comment by Undefined-Variables** [comment]
Hey @pH-7 

I don't want you to think I am picking on the software, that was not the point of opening this issue, I have a lot of love for the software. This software has become my life over the past two years and I think open source is the best way to develop software but I think we could really take this to the next level with some planning between ourselves and by implementing some fairly  simple changes/updates.

1. Messages 
The current system is very outdated, what I mean by this is every app I can think of approaches user communication in the style of a back and forth conversation, so user A says "Hi how are you today?" then user B replies with "Hi I am doing well, how are you?"  When user A reads the new message he can see his original message the reply and has a text box where he can add to the conversation. 

Why? because when you are talking to 20 people and you get a new message you have no idea what you said or what he said and you reply with a message that doesn't really apply to the past communication between you. The end user (the site member) would feel much more connected and involved in the website if they could see their message history! 

2. Instant Messenger
Currently its very basic and its hit and miss, sometimes the box pops up sometimes it doesn't and this gives the impression your message was seen but ignored. This causes an atmosphere between users because they feel people are rude and are being ignorant. On a dating or social site we want to promote a happy and friendly feeling. Being ignored just annoys the users.

I have been working with ajax and a lot of JavaScript lately and I am having some real fun with it so I may take the task of revamping the Instant messenger in the near future. Personally I think FaceBook have the right idea, showing who is online and in a neat easy to access list. This is the only thing I like about FB as I don't like the limelight and attention it brings. 
  
3. Friends 
When you add a friend they are instantly approved from their own point of view. The other user has the choice to approve or delete but the request has already been approved, the variables $is_pending_friend and $is_approved_friend are not holding the correct values and make no sense.

The query for mutual friends also throws errors after it gets to a certain size as I have explained here #793

4. Install Script
In step 4 (config_system)...

````
/********************* STEP 4 *********************/
    public function config_system()
    {
        global $LANG;

        if (!empty($_SESSION['step3']) && is_file(PH7_ROOT_PUBLIC . '_constants.php')) {
            session_regenerate_id(true);

            if (empty($_SESSION['val'])) {
                $_SESSION['db']['type_name'] = Db::DBMS_MYSQL_NAME;
                $_SESSION['db']['type'] = Db::DSN_MYSQL_PREFIX;

                $_SESSION['db']['hostname'] = DbDefaultConfig::HOSTNAME;
                $_SESSION['db']['username'] = DbDefaultConfig::USERNAME;
                $_SESSION['db']['name'] = DbDefaultConfig::NAME;
                $_SESSION['db']['prefix'] = DbDefaultConfig::PREFIX;
                $_SESSION['db']['port'] = DbDefaultConfig::PORT;
                $_SESSION['db']['charset'] = DbDefaultConfig::CHARSET;

                $_SESSION['val']['bug_report_email'] = '';
                $_SESSION['val']['ffmpeg_path'] = ffmpeg_path();
            }

            if ($_SERVER['REQUEST_METHOD'] === 'POST' && !empty($_POST['config_system_submit'])) {
                if (filled_out($_POST)) {
                    foreach ($_POST as $sKey => $sVal) {
                        $_SESSION['db'][str_replace('db_', '', $sKey)] = trim($sVal);
                    }

                    $_SESSION['val']['bug_report_email'] = trim($_POST['bug_report_email']);
                    $_SESSION['val']['ffmpeg_path'] = trim($_POST['ffmpeg_path']);

                    if (validate_email($_SESSION['val']['bug_report_email'])) {
                        try {
                            require_once PH7_ROOT_INSTALL . 'inc/_db_connect.inc.php';
                            @require_once PH7_ROOT_PUBLIC . '_constants.php';
                            @require_once PH7_PATH_APP . 'configs/constants.php';

                            // Config File
                            @chmod(PH7_PATH_APP_CONFIG, 0777);
                            $sConfigContent = file_get_contents(PH7_ROOT_INSTALL . 'data/configs/config.ini');

                            $sConfigContent = str_replace('%bug_report_email%', $_SESSION['val']['bug_report_email'], $sConfigContent);
                            $sConfigContent = str_replace('%ffmpeg_path%', clean_string($_SESSION['val']['ffmpeg_path']), $sConfigContent);

                            $sConfigContent = str_replace('%db_type_name%', $_SESSION['db']['type_name'], $sConfigContent);
                            $sConfigContent = str_replace('%db_type%', $_SESSION['db']['type'], $sConfigContent);
                            $sConfigContent = str_replace('%db_hostname%', $_SESSION['db']['hostname'], $sConfigContent);
                            $sConfigContent = str_replace('%db_username%', clean_string($_SESSION['db']['username']), $sConfigContent);
                            $sConfigContent = str_replace('%db_password%', clean_string($_SESSION['db']['password']), $sConfigContent);
                            $sConfigContent = str_replace('%db_name%', clean_string($_SESSION['db']['name']), $sConfigContent);
                            $sConfigContent = str_replace('%db_prefix%', clean_string($_SESSION['db']['prefix']), $sConfigContent);
                            $sConfigContent = str_replace('%db_charset%', $_SESSION['db']['charset'], $sConfigContent);
                            $sConfigContent = str_replace('%db_port%', $_SESSION['db']['port'], $sConfigContent);

                            $sConfigContent = str_replace('%private_key%', generate_hash(40), $sConfigContent);
                            $sConfigContent = str_replace('%rand_id%', generate_hash(5), $sConfigContent);

                            if (!@file_put_contents(PH7_PATH_APP_CONFIG . 'config.ini', $sConfigContent)) {
                                $aErrors[] = $LANG['no_app_config_writable'];
                            } else {
                                if (!(
                                    $DB->getAttribute(\PDO::ATTR_DRIVER_NAME) === Db::DSN_MYSQL_PREFIX &&
                                    version_compare($DB->getAttribute(\PDO::ATTR_SERVER_VERSION), PH7_REQUIRED_SQL_VERSION, '>='))
                                ) {
                                    $aErrors[] = $LANG['require_mysql_version'];
                                } else {
                                    ignore_user_abort(true);

                                    $aDumps = [
                                        /** Game **/
                                        // We need to install the Game before the "Core SQL" for foreign key reasons
                                        'pH7_SchemaGame',
                                        'pH7_DataGame',
                                        /** Core (main SQL schema/data) **/
                                        'pH7_Core'
                                    ];

                                    for ($iFileKey = 0, $iCount = count($aDumps); $iFileKey < $iCount; $iFileKey++) {
                                        exec_query_file(
                                            $DB,
                                            PH7_ROOT_INSTALL . 'data/sql/' . $_SESSION['db']['type_name'] . '/' . $aDumps[$iFileKey] . '.sql'
                                        );
                                    }

                                    // We finalise it by setting the correct permission to the config files
                                    $this->chmodConfigFiles();

                                    $_SESSION['step4'] = 1;
                                    unset($_SESSION['val']);

                                    redirect(PH7_URL_SLUG_INSTALL . 'config_site');
                                }
                            }
                        } catch (\PDOException $oE) {
                            $aErrors[] = $LANG['database_error'] . escape($oE->getMessage());
                        }
                    } else {
                        $aErrors[] = $LANG['bad_email'];
                    }
                } else {
                    $aErrors[] = $LANG['all_fields_mandatory'];
                }
            }
        } else {
            redirect(PH7_URL_SLUG_INSTALL . 'config_path');
        }

        // Assign the sample DB config values to the template
        $this->oView->assign('def_db_hostname', DbDefaultConfig::HOSTNAME);
        $this->oView->assign('def_db_username', DbDefaultConfig::USERNAME);
        $this->oView->assign('def_db_name', DbDefaultConfig::NAME);
        $this->oView->assign('def_db_prefix', DbDefaultConfig::PREFIX);
        $this->oView->assign('def_db_port', DbDefaultConfig::PORT);
        $this->oView->assign('def_db_charset', DbDefaultConfig::CHARSET);

        $this->oView->assign('sept_number', 4);
        $this->oView->assign('errors', @$aErrors);
        unset($aErrors);

        $this->oView->display('config_system.tpl');
    }
```` 

Do you see how we insert the tables in to the database

````

                                    $aDumps = [
                                        /** Game **/
                                        // We need to install the Game before the "Core SQL" for foreign key reasons
                                        'pH7_SchemaGame',
                                        'pH7_DataGame',
                                        /** Core (main SQL schema/data) **/
                                        'pH7_Core'
                                    ];
````

Before we validate and return the errors? So if you have an error in one of the fields you are redirected to the start again but the tables are already inserted so what happens is step 5 gets skipped. This means users believe the install was successful but have no admin access cos they didn't set the admin account up in step 5.

5. I have some other concerns I do not want to discuss publicly yet and I am still investigating. I will send you an email once I have a full understanding and would appreciate it if you read it and reply before I open a public issue. I am willing to give my time to help improve this software but others who use it have to be contributing to it to. There is no joy if one person has to do everything. 

There are tons of people using this and who will read this and have made improvements to their own copy that is running their websites but they just take take take, all the updates, additions and hard work of a small few. Its time that you give something back to the very thing that has helped you and you dream or idea to become a reality!

**Comment by polynamaude** [comment]
@Undefined-Variables If you want people to get involved, you have to make it show that this ain't like a "one man" project.
I mean, I have never saw a request for help, I saw asking what people do with the software for the first time last week.
Look at other project who get feedback from community...

**Comment by Undefined-Variables** [comment]
@polynamaude I agree the question "What do you do with the software" is a good question to ask to make people feel part of something. 

Its great software as it stands but it could be AWESOME if the community all got more involved, just look at the life changing projects that are happening all over GitHub... This is mostly due to teamwork and great project leaders. We could make this truly magical if we all put our share in!

**Comment by Undefined-Variables** [comment]
Totally forgot about this I did as an example of messages in the way I described above.(I did this a few months back)

https://undefined-variables.com/ph7cms/

There are demo user 1 and user 2 and password on this page, use these to seea user with message history!

Im looking at the IM module and there is a lot of things done that could be more efficiently like sterilize() why not json_encode? its faster...  just one example there is a lot more like setJsonContent() can be done in better way and calling $_SESSION over and over again is avoidable... Think I can make it better with a small amount of work... leave it with me...

**Comment by pH-7** [comment]
> Hey @pH-7
> 
> I don't want you to think I am picking on the software, that was not the point of opening this issue, I have a lot of love for the software. This software has become my life over the past two years and I think open source is the best way to develop software but I think we could really take this to the next level with some planning between ourselves and by implementing some fairly simple changes/updates.
> 
>     1. Messages
>        The current system is very outdated, what I mean by this is every app I can think of approaches user communication in the style of a back and forth conversation, so user A says "Hi how are you today?" then user B replies with "Hi I am doing well, how are you?"  When user A reads the new message he can see his original message the reply and has a text box where he can add to the conversation. Why? because when you are talking to 20 people and you get a new message you have no idea what you said or what he said and you reply with a message that doesn't really apply to the past communication between you. The end user (the site member) would feel much more connected and involved in the website if they could see their message history!
> 
>     2. Instant Messenger
>        Currently its very basic and its hit and miss, sometimes the box pops up sometimes it doesn't and this gives the impression your message was seen but ignored. This causes an atmosphere between users because they feel people are rude and are being ignorant. On a dating or social site we want to promote a happy and friendly feeling. Being ignored just annoys the users.
>        I have been working with ajax and a lot of JavaScript lately and I am having some real fun with it so I may take the task of revamping the Instant messenger in the near future. Personally I think FaceBook have the right idea, showing who is online and in a neat easy to access list. This is the only thing I like about FB as I don't like the limelight and attention it brings.
> 
>     3. Friends
>        When you add a friend they are instantly approved from their own point of view. The other user has the choice to approve or delete but the request has already been approved, the variables $is_pending_friend and $is_approved_friend are not holding the correct values and make no sense. The query for mutual friends also throws errors after it gets to a certain size as I have explained here #793
> 
>     4. Install Script
>        In step 4 (config_system)...
> 
> 
> ```
> /********************* STEP 4 *********************/
>     public function config_system()
>     {
>         global $LANG;
> 
>         if (!empty($_SESSION['step3']) && is_file(PH7_ROOT_PUBLIC . '_constants.php')) {
>             session_regenerate_id(true);
> 
>             if (empty($_SESSION['val'])) {
>                 $_SESSION['db']['type_name'] = Db::DBMS_MYSQL_NAME;
>                 $_SESSION['db']['type'] = Db::DSN_MYSQL_PREFIX;
> 
>                 $_SESSION['db']['hostname'] = DbDefaultConfig::HOSTNAME;
>                 $_SESSION['db']['username'] = DbDefaultConfig::USERNAME;
>                 $_SESSION['db']['name'] = DbDefaultConfig::NAME;
>                 $_SESSION['db']['prefix'] = DbDefaultConfig::PREFIX;
>                 $_SESSION['db']['port'] = DbDefaultConfig::PORT;
>                 $_SESSION['db']['charset'] = DbDefaultConfig::CHARSET;
> 
>                 $_SESSION['val']['bug_report_email'] = '';
>                 $_SESSION['val']['ffmpeg_path'] = ffmpeg_path();
>             }
> 
>             if ($_SERVER['REQUEST_METHOD'] === 'POST' && !empty($_POST['config_system_submit'])) {
>                 if (filled_out($_POST)) {
>                     foreach ($_POST as $sKey => $sVal) {
>                         $_SESSION['db'][str_replace('db_', '', $sKey)] = trim($sVal);
>                     }
> 
>                     $_SESSION['val']['bug_report_email'] = trim($_POST['bug_report_email']);
>                     $_SESSION['val']['ffmpeg_path'] = trim($_POST['ffmpeg_path']);
> 
>                     if (validate_email($_SESSION['val']['bug_report_email'])) {
>                         try {
>                             require_once PH7_ROOT_INSTALL . 'inc/_db_connect.inc.php';
>                             @require_once PH7_ROOT_PUBLIC . '_constants.php';
>                             @require_once PH7_PATH_APP . 'configs/constants.php';
> 
>                             // Config File
>                             @chmod(PH7_PATH_APP_CONFIG, 0777);
>                             $sConfigContent = file_get_contents(PH7_ROOT_INSTALL . 'data/configs/config.ini');
> 
>                             $sConfigContent = str_replace('%bug_report_email%', $_SESSION['val']['bug_report_email'], $sConfigContent);
>                             $sConfigContent = str_replace('%ffmpeg_path%', clean_string($_SESSION['val']['ffmpeg_path']), $sConfigContent);
> 
>                             $sConfigContent = str_replace('%db_type_name%', $_SESSION['db']['type_name'], $sConfigContent);
>                             $sConfigContent = str_replace('%db_type%', $_SESSION['db']['type'], $sConfigContent);
>                             $sConfigContent = str_replace('%db_hostname%', $_SESSION['db']['hostname'], $sConfigContent);
>                             $sConfigContent = str_replace('%db_username%', clean_string($_SESSION['db']['username']), $sConfigContent);
>                             $sConfigContent = str_replace('%db_password%', clean_string($_SESSION['db']['password']), $sConfigContent);
>                             $sConfigContent = str_replace('%db_name%', clean_string($_SESSION['db']['name']), $sConfigContent);
>                             $sConfigContent = str_replace('%db_prefix%', clean_string($_SESSION['db']['prefix']), $sConfigContent);
>                             $sConfigContent = str_replace('%db_charset%', $_SESSION['db']['charset'], $sConfigContent);
>                             $sConfigContent = str_replace('%db_port%', $_SESSION['db']['port'], $sConfigContent);
> 
>                             $sConfigContent = str_replace('%private_key%', generate_hash(40), $sConfigContent);
>                             $sConfigContent = str_replace('%rand_id%', generate_hash(5), $sConfigContent);
> 
>                             if (!@file_put_contents(PH7_PATH_APP_CONFIG . 'config.ini', $sConfigContent)) {
>                                 $aErrors[] = $LANG['no_app_config_writable'];
>                             } else {
>                                 if (!(
>                                     $DB->getAttribute(\PDO::ATTR_DRIVER_NAME) === Db::DSN_MYSQL_PREFIX &&
>                                     version_compare($DB->getAttribute(\PDO::ATTR_SERVER_VERSION), PH7_REQUIRED_SQL_VERSION, '>='))
>                                 ) {
>                                     $aErrors[] = $LANG['require_mysql_version'];
>                                 } else {
>                                     ignore_user_abort(true);
> 
>                                     $aDumps = [
>                                         /** Game **/
>                                         // We need to install the Game before the "Core SQL" for foreign key reasons
>                                         'pH7_SchemaGame',
>                                         'pH7_DataGame',
>                                         /** Core (main SQL schema/data) **/
>                                         'pH7_Core'
>                                     ];
> 
>                                     for ($iFileKey = 0, $iCount = count($aDumps); $iFileKey < $iCount; $iFileKey++) {
>                                         exec_query_file(
>                                             $DB,
>                                             PH7_ROOT_INSTALL . 'data/sql/' . $_SESSION['db']['type_name'] . '/' . $aDumps[$iFileKey] . '.sql'
>                                         );
>                                     }
> 
>                                     // We finalise it by setting the correct permission to the config files
>                                     $this->chmodConfigFiles();
> 
>                                     $_SESSION['step4'] = 1;
>                                     unset($_SESSION['val']);
> 
>                                     redirect(PH7_URL_SLUG_INSTALL . 'config_site');
>                                 }
>                             }
>                         } catch (\PDOException $oE) {
>                             $aErrors[] = $LANG['database_error'] . escape($oE->getMessage());
>                         }
>                     } else {
>                         $aErrors[] = $LANG['bad_email'];
>                     }
>                 } else {
>                     $aErrors[] = $LANG['all_fields_mandatory'];
>                 }
>             }
>         } else {
>             redirect(PH7_URL_SLUG_INSTALL . 'config_path');
>         }
> 
>         // Assign the sample DB config values to the template
>         $this->oView->assign('def_db_hostname', DbDefaultConfig::HOSTNAME);
>         $this->oView->assign('def_db_username', DbDefaultConfig::USERNAME);
>         $this->oView->assign('def_db_name', DbDefaultConfig::NAME);
>         $this->oView->assign('def_db_prefix', DbDefaultConfig::PREFIX);
>         $this->oView->assign('def_db_port', DbDefaultConfig::PORT);
>         $this->oView->assign('def_db_charset', DbDefaultConfig::CHARSET);
> 
>         $this->oView->assign('sept_number', 4);
>         $this->oView->assign('errors', @$aErrors);
>         unset($aErrors);
> 
>         $this->oView->display('config_system.tpl');
>     }
> ```
> 
> Do you see how we insert the tables in to the database
> 
> ```
> 
>                                     $aDumps = [
>                                         /** Game **/
>                                         // We need to install the Game before the "Core SQL" for foreign key reasons
>                                         'pH7_SchemaGame',
>                                         'pH7_DataGame',
>                                         /** Core (main SQL schema/data) **/
>                                         'pH7_Core'
>                                     ];
> ```
> 
> Before we validate and return the errors? So if you have an error in one of the fields you are redirected to the start again but the tables are already inserted so what happens is step 5 gets skipped. This means users believe the install was successful but have no admin access cos they didn't set the admin account up in step 5.
> 
>     1. I have some other concerns I do not want to discuss publicly yet and I am still investigating. I will send you an email once I have a full understanding and would appreciate it if you read it and reply before I open a public issue. I am willing to give my time to help improve this software but others who use it have to be contributing to it to. There is no joy if one person has to do everything. There are tons of people using this and who will read this and have made improvements to their own copy that is running their websites but they just take take take, all the updates, additions and hard work of a small few. Its time that you give something back to the very thing that has helped you and you dream or idea to become a reality!

This is definitely so useful @Undefined-Variables 🎉!! Thank you so much for those details. I will probably split each point into an issue, to make the workflow easier.

**Comment by pH-7** [comment]
> I have some other concerns I do not want to discuss publicly yet and I am still investigating. I will send you an email once I have a full understanding and would appreciate it if you read it and reply before I open a public issue. I am willing to give my time to help improve this software but others who use it have to be contributing to it to. There is no joy if one person has to do everything. There are tons of people using this and who will read this and have made improvements to their own copy that is running their websites but they just take take take, all the updates, additions and hard work of a small few. Its time that you give something back to the very thing that has helped you and you dream or idea to become a reality!

Awesome 🤗 Yes, totally agree!

**Comment by pH-7** [comment]
> Its great software as it stands but it could be AWESOME if the community all got more involved, just look at the life changing projects that are happening all over GitHub... This is mostly due to teamwork and great project leaders. We could make this truly magical if we all put our share in!

So glad to have you in the community anyway! 😊 You are definitely a great element!

**Comment by pH-7** [comment]
In the meantime, like I said in a previous message earlier last week in this thread, I will already try to release the stable v16 version sometime this week..

**Comment by Undefined-Variables** [comment]
I have been hard at work on the messenger and I will be writing a version for the software. Its just a joy to use.... Here is a couple of screen shots to give you an idea!

![444](https://user-images.githubusercontent.com/52648055/89737874-74da2d00-da6c-11ea-9f49-454f3b1e6bf2.PNG)
![2122](https://user-images.githubusercontent.com/52648055/89737875-7572c380-da6c-11ea-8ba7-49d36f300184.PNG)
![11](https://user-images.githubusercontent.com/52648055/89737876-760b5a00-da6c-11ea-9160-9dbcf1e5f38f.PNG)
![66666](https://user-images.githubusercontent.com/52648055/89737877-760b5a00-da6c-11ea-8e53-47d4c30e9de5.PNG)
![55555](https://user-images.githubusercontent.com/52648055/89737878-76a3f080-da6c-11ea-8219-cbd97dd14bcb.PNG)

**Comment by Undefined-Variables** [comment]
This is how a messenger should look and function in the modern age!  It minimises to the little green box and is reliable and invites you to chat cos you can see who is online! I'm really getting my teeth in to js and ajax in particular. Having lots of fun so that is what matters!

**Comment by polynamaude** [comment]
@Undefined-Variables I just love the pictures you used. Looks like the good old "Lego" from when I was a kid and build stuff (also blew it off as my dad was in the military)...

**Comment by breisig** [comment]
Wow, this looks really good. Would love to see this integrated.

**Comment by Undefined-Variables** [comment]
@polynamaude yeah me too, the choice of picture was completely random but it also takes me back to my childhood. I have a sister who is 18 months younger than me and it was her job to find me the pieces I needed 😁 and all these years later she still reminds me everytime she sees Lego! 

@breisig I will write a similar version to be integrated in to the software. Just give me a little time. I have a lot on this week but I will squeezee some time in for this. The one you see was written while I was watching fast and furious 7 with my daughter today.

**Comment by Undefined-Variables** [comment]
@pH-7 Check your gmail... Cheers mate.

**Comment by polynamaude** [comment]
@Undefined-Variables Don't tell people to check their mail ;-) You don't know their agenda, maybe he already saw your mail but hadn't time to give it a answer.

**Comment by Undefined-Variables** [comment]
@polynamaude I just sent it but got an auto so thought id tell him I had sent one. I was not rushing  haha!

**Comment by pH-7** [comment]
> I have been hard at work on the messenger and I will be writing a version for the software. Its just a joy to use.... Here is a couple of screen shots to give you an idea!
> 
> 
> 
> ![444](https://user-images.githubusercontent.com/52648055/89737874-74da2d00-da6c-11ea-9f49-454f3b1e6bf2.PNG)
> 
> ![2122](https://user-images.githubusercontent.com/52648055/89737875-7572c380-da6c-11ea-8ba7-49d36f300184.PNG)
> 
> ![11](https://user-images.githubusercontent.com/52648055/89737876-760b5a00-da6c-11ea-9160-9dbcf1e5f38f.PNG)
> 
> ![66666](https://user-images.githubusercontent.com/52648055/89737877-760b5a00-da6c-11ea-8e53-47d4c30e9de5.PNG)
> 
> ![55555](https://user-images.githubusercontent.com/52648055/89737878-76a3f080-da6c-11ea-8219-cbd97dd14bcb.PNG)
> 
> 

Wow! 🤩 Amazing work so far @Undefined-Variables!

**Comment by pH-7** [comment]
> @pH-7 Check your gmail... Cheers mate.

What's the email you sent to?

**Comment by pH-7** [comment]
> @pH-7 Check your gmail... Cheers mate.

Was it sent at ph7software {AT] gmail {D0T} com?
If so, I will chack this mailbox by now 😊

**Comment by Undefined-Variables** [comment]
@pH-7 have you got a minute? I am stuck on a problem and I am unsure how to word what I want to do so its hard to research something if you don't know what its referred to. This is part of the instant messenger update so it will enhance the software...

So I use a true or false, 0 or 1 to determine if the user is online/active like

    	$this->oMessengerModel->update($id, array('online'=>'1'));

I want to set a timeout on this so after 5 minutes or 300 Seconds it will go

        $this->oUserModel->update($id, array('online'=>'0'));

Automatically but all I can find on PHP timeout is max execute script timeout and this is not what I am looking for. Do I use a function and loop until the time expires or how is the best way to do this and what is doing this called? I can learn if I can find...

**Comment by Undefined-Variables** [comment]
Just to elaborate before someone gives an obvious answer, yes I understand that PHP is executing server side and I know about sleep() but this is blocking, so it pauses the execution of the script... This is not what I am trying to do. 

I know this can be done cos @pH-7 has used this on user timeout but I don't see how he has managed to do it so that is what I'd like to know in detail please 💯

**Comment by polynamaude** [comment]
@Undefined-Variables Again, I'll go with reminding you there's reason why learning has a path to follow...
Because there's a thing called archeology of knowledge, level of learning...

What you are trying to get done ain't something do be done in the script itself... or at least not in the main call... You can't make your PHP script wait in a loop because this would first hang your release to Apache call (as PHP is in interaction with your web server to process the web page and serve it back to Apache).
This is what we call Asynchronous jobs... What you can do is simply timestamp in a database table and validate this when needed if it was exceeded. For example when someone want to know who's online, then you look who haven't "punched" in the table that he's still there. So you create a JS function, that will call pH7CMS every x seconds, saying "Hi, I'm still alive". and this function update the table... So you'll probably need to either use a service worker or something similar in JS, plus write a back end interface...

Using something akward as ReactPHP wouldn't be the solution... Unless you want to do it dirty because the other solution seem hard to understand...

There's a place for everything and what you are writing here is in the "front end", I mean, it's the client that call the server and say "Hi ! I'm alive" so it shall be done in a front-end language, that's Javascript... PHP is a back-end processing language.... Evolved much from Personal Home Page like it was at first !

**Comment by polynamaude** [comment]
@Undefined-Variables And this is why there's a boring part of learning different type of building block for solution, including a nice book on design pattern and software architectures...

Because now I'm gonna ask the "Kill question"... Are you just writing software as you write a note to a friend or did you take time to plan what you gonna do, with list of variables, graph of process / decision and all that boring planning ?

Have you planned what to do if you have more than one connexion from the same user ? And how to manage the logoff ? You just "let it expire" or do you force expiry by a function linked to the logout procedure ? ... ?

**Comment by Undefined-Variables** [comment]
No I understand how this works in JavaScript and can do this using JQuery and asynchronously too, I have done lots of work lately using asynchronous features. The wheel for example, it spins and I was facing the problem that with no refreshing of the page the number of spins would never change, from the users point of view. I got passed this by writing a script that asynchronously queries the database and updates the values without refreshing the page.

This is why I tried to explain I know sleep is not the correct way of thinking as would just stall the script. 

I know its done in the software already as you set the timeout before the user seems offline. So ideally I want to use the same method to do the same thing as is already done. I know roughly how this works before I have looked at the script.

Yeah as far as the logout goes if a user was to logout properly using the sites logout then the session is destroyed that the chat is built upon so yes the logout would have the desired effect but I know as well as you that not many users logout they just close the browser or tab... This is an issue because with no timeout everyone is online 100% of the time. which is why I started looking at ways to do this.

I am not shy of learning but if you dont know what something is called its very hard to search for the answer...

**Comment by Undefined-Variables** [comment]
Found where its done and how... Super simple... There is nothing to hard or to challenging... The more it makes me think the better...

````
function updateActivity() {
    $.get(pH7Url.base + 'user/asset/ajax/setActivity');
    setTimeout('updateActivity()', 10000)
}
updateActivity();
````

That is simple as it gets :)

**Comment by Undefined-Variables** [comment]
> There's a place for everything and what you are writing here is in the "front end", I mean, it's the client that call the server and say "Hi ! I'm alive" so it shall be done in a front-end language, that's Javascript... PHP is a back-end processing language.... Evolved much from Personal Home Page like it was at first !

Yes I know and do agree, its not about how hard it is to do its about feeling confident in the code I am writing and knowing how it will react under load and JS im never as happy and confident with it. I am still learning JS. 

PHP is like writing in English with a pen and paper... It just feels right you know....

**Comment by Undefined-Variables** [comment]
> @Undefined-Variables And this is why there's a boring part of learning different type of building block for solution, including a nice book on design pattern and software architectures...
> 
I have read a few to be fair, but I don't learn by reading I learn by doing... 

> Because now I'm gonna ask the "Kill question"... Are you just writing software as you write a note to a friend or did you take time to plan what you gonna do, with list of variables, graph of process / decision and all that boring planning ?
> 
Come on @polynamaude I plan everything I do, first on scrap paper, then the idea gets more focussed, I make a clear plan and outline the tables, site functionality and then code a mock to give my business partner an idea of my plan, if he agrees we move forward if not we start from square one again. 

> Have you planned what to do if you have more than one connexion from the same user ? And how to manage the logoff ? You just "let it expire" or do you force expiry by a function linked to the logout procedure ? ... ?
>
Yes a second connection to the same account will result in the first users session being destroyed on login of the second instance. An email warning them there account has been accessed from another location should be sent just incase its not them.

**Comment by polynamaude** [comment]
@Undefined-Variables I think you're pushing it a bit far here...
Sending the user a email only because he logged in twice in a row.. Okay so I can't talk with someone using my mobile phone and at the same time (or within a short time) use my computer... If I forget to log out, I get a email that now is scaring me that maybe my account was hacked... That make a bit non-sense...

You said you didn't want to write JS but now what you are putting a front is some Javascript code...

If you'd plan before starting this Messenger add-on, you wouldn't ask the question now but it would have surfaced earlier...

**Comment by Undefined-Variables** [comment]
@polynamaude every time I sign in to gmail from a different device than the previous one I get 20 Question, texts sent, email code and I have to okay it on the last device I used... That is going a little far right? I mean telling someone they have signed in multiple times on a dating site might be appreciated, it could be the other half at home spying on them haha!

Do you read my words mate? I said I feel more comfortable when working with PHP not that I don't like JS cos I do there is just things I am still learning and understanding the weird parts like any language come with experience!    

This did cross my mind earlier your right but I thought this was in place which it is... There will always be something you overlook that is just part of being human!

**Comment by Undefined-Variables** [comment]
> If you'd plan before starting this Messenger add-on, you wouldn't ask the question now but it would have surfaced earlier...

This is playing on my mind... Has to be said... 

If you criticise people when they are giving their time and trying to improve the software you make them feel like "What is the point" maybe this approach towards people who try and improve things is just driving them away! You have to feel happy, valued and at ease where ever you spend time and that is not the vibe i get here! 

Why always so negative? I think encouragement would do so much more! Feeling fed up tbh! I only been back a day!

**Comment by polynamaude** [comment]
@Undefined-Variables What I said was very simple...
It's part of your application that goes in the front-end.

So if you knew that it's front end then why do you ask
**Automatically but all I can find on PHP timeout is max execute script timeout and this is not what I am looking for. Do I use a function and loop until the time expires or how is the best way to do this and what is doing this called? I can learn if I can find...**

Those are your own words...

Also, maybe you gonna repeat me... I don't learn by reading but doing. But that's how we all do. So that's why there's something called "Practial exercise" when you take tutorials and courses. Even books that teach you programming will guide you thru project. You shall try this on either OpenClassroom, Udacity or something similar. Not so expensive, you can get around for at most 20$ a month. No excuse that your business partner wouldn't pay for this ;-)

**Comment by Undefined-Variables** [comment]
@polynamaude your right, I should know better than to think out load here... I have a funny route of thought sometimes and you have been taught to think the same problem though in a very structured way.

Don't think I let what you say go by without taking note and next time I will think the same. Tbf @polynamaude you have taught me a lot and you probably don't realise it. Sorry had a full on week with my 17yr old daughter and think I'm not as focused as normal but I have a backlog of things to finish.

**Comment by Undefined-Variables** [comment]
@polynamaude I won't let him pay for anything. He runs the site and development is my area. I want to complete the site myself, its been 2 1/2 yrs and I am almost done.

**Comment by Undefined-Variables** [comment]
@pH-7 How do I start a new Wiki page again? Its not letting me?
